src/ray/common/status_or.h

```diff
@@ -0,0 +1,257 @@
+      : data_(std::forward<Args>(args)...) {}
+
+  template <typename U>
+  StatusOr(const StatusOr<U> &status_or) {
+    if (status_or.ok()) {
+      data_ = status_or.value();
+      status_ = Status::OK();
+    } else {
+      status_ = status_or.status();
+    }
+  }
+
+  template <typename U>
+  StatusOr(StatusOr<U> &&status_or) {
+    if (status_or.ok()) {
+      data_ = std::move(status_or).value();
+      status_ = Status::OK();
+    } else {
+      status_ = std::move(status_or).status();
+    }
+  }
+
+  StatusOr(const StatusOr &) = default;
+  StatusOr &operator=(const StatusOr &) = default;
+
+  StatusOr(StatusOr &&) noexcept(std::is_nothrow_move_constructible_v<T>) = default;
+  StatusOr &operator=(StatusOr &&) noexcept(std::is_nothrow_move_assignable_v<T>) =
+      default;
+
+  ~StatusOr() noexcept(std::is_nothrow_destructible_v<T>) = default;
+
+  // Returns whether or not this `ray::StatusOr<T>` holds a `T` value. This
+  // member function is analogous to `ray::Status::ok()` and should be used
+  // similarly to check the status of return values.
+  //
+  // Example:
+  //
+  // StatusOr<Foo> result = DoBigCalculationThatCouldFail();
+  // if (result.ok()) {
+  //    // Handle result
+  // else {
+  //    // Handle error
+  // }
+  bool ok() const { return status_.ok(); }
+  bool has_value() const { return ok(); }
+  explicit operator bool() const { return ok(); }
+
+  template <typename U>
+  T value_or(U &&u) {
+    return ok() ? get() : T{std::forward<U>(u)};
+  }
+
+  ABSL_MUST_USE_RESULT StatusCode code() const { return status_.code(); }
+
+  ABSL_MUST_USE_RESULT std::string message() const { return status_.message(); }
+
+  // Returns a reference to the current `ray::Status` contained within the
+  // `ray::StatusOr<T>`. If `ray::StatusOr<T>` contains a `T`, then this
+  // function returns `ray::Ok()`.
+  ABSL_MUST_USE_RESULT const Status &status() const & { return status_; }
+  ABSL_MUST_USE_RESULT Status status() && {
+    Status new_status = std::move(status_);
+    return new_status;
+  }
+
+  // Apply the functor [f] if `this->ok() == true`, otherwise return the error status
+  // contained.
+  template <typename F>
+  auto and_then(F &&f) &;
+  template <typename F>
+  auto and_then(F &&f) const &;
+  template <typename F>
+  auto and_then(F &&f) &&;
+
+  // Returns a reference to the current value.
+  //
+  // REQUIRES: `this->ok() == true`, otherwise the behavior is undefined.
+  //
+  // Use `this->ok()` to verify that there is a current value within the
+  // `ray::StatusOr<T>`. Alternatively, see the `value()` member function for a
+  // similar API that guarantees crashing or throwing an exception if there is
+  // no current value.
+  T &operator*() & { return get(); }
+  const T &operator*() const & { return get(); }
+  T &&operator*() && { return std::move(get()); }
+
+  // Returns a pointer to the current value.
+  //
+  // REQUIRES: `this->ok() == true`, otherwise the behavior is undefined.
```

Comment:
instead of UB, do we want to CHECK fail here?

Reply:
I'm following the abseil interface for accessing part.
https://github.com/abseil/abseil-cpp/blob/fcc8630eede5498b48b45b99e4eaa1406d6657e9/absl/status/statusor.h#L501-L514

The reason for not checking is performance, since users should check OK before accessing.
```
if (statusor.ok()) {
  value = *statusor;
}
```

Reply:
usually these api's like optional have something like a .value() that does a check and a throws an exception, and then the *(optional), gives you the value without the check, might make sense to stick to that established concept